web/web/models/alert.py

```diff
@@ -1,28 +1,235 @@
+                         nullable=False)
+
+    description = Column(db.Text,
+                         nullable=False)
+
+    status = Column(db.Enum(Status),
+                    nullable=False)
+
+    context = Column(db.Enum(ContextType),
+                     nullable=False)
+
+    context_id = Column(db.String(64),
+                        nullable=False)
+
+    resolution = Column(db.Text,
+                        nullable=False)
+    
+    updated_at = Column(TIMESTAMP, 
+                        nullable=False,
+                        server_default=text('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'))
+
+    created_at = Column(TIMESTAMP,
+                        nullable=False,
+                        server_default=func.now())
+
+    # Methods
+    def __repr__(self):
+        return f'<Alert alert_id={self.alert_id} alert_type_id={self.alert_type_id} created_at={self.created_at}>'
+
+    def get_filter_type(self):
+        '''Returns filter type as a string for alert frontend set-up'''
+
+        if self.alert_type.name.value == 'Import Capacity' \
+            or self.alert_type.name.value == 'Export Capacity' \
+                or self.alert_type.name.value == 'Load Yellow' \
+                    or self.alert_type.name.value == 'Load Red':
+            return 'Capacity bounds'
+        
+        if self.alert_type.name.value == 'Price Yellow' \
+            or self.alert_type.name.value == 'Price Red' \
+                or self.alert_type.name.value == 'Price':
+            return 'Price alerts'
+        
+        if self.alert_type.name.value == 'Telecomm':
+            return 'Telecomm alerts'
+        
+        if self.alert_type.name.value == 'Resource':
+            return 'Resource depletion'
+        
+        if self.alert_type.name.value == 'Peak Event':
+            return 'Peak event'
+
+# Relationship declared on other table (dependency on table import for event listener)
+AlertType.alerts = relationship('Alert',
+                                 backref=db.backref('alert_type'))
+
+# Event listener for email notifications
+@event.listens_for(Alert, 'after_insert')
+def after_insert(mapper, connection, target):
+    '''Sends email after a new alert event is inserted into database'''
+
+    # Query all notifications that are active for the alert type
+    notifications = Notification.query \
+                                .filter(Notification.alert_type_id==target.alert_type_id, Notification.is_active==True) \
+                                .all()
+
+    # Checks if there are no notifications for early exit
+    if notifications == []:
+        return None
+
+    # Since after_insert event is triggered before relationships are configured, alert_type relationship needs to be declared
+    alert_type = AlertType.query \
+                          .filter_by(alert_type_id=target.alert_type_id) \
+                          .first()
+    target.alert_type = alert_type
```

Comment:
Does assigning it to target.alert_type do anything?

Reply:
Yes. Without establishing the relationship explicitly here, the relationship does not yet exist because the way "after insert" is triggered is before relationships are mapped.